[MGDOBR-403] Improve deployment scripts for local dev

radtriste opened the pull request:
[MGDOBR-403](https://issues.redhat.com/browse/MGDOBR-403) 

See also https://github.com/radtriste/sandbox/blob/deploy_script/dev/README.md

```bash
$ ./dev/bin/deploy.sh -h
Usage: deploy.sh [OPTIONS]

Options:
  -l                  Use images built locally (they will be built automatically)
  -s                  Start the cluster
  -a                  Deploy all resources on the cluster

Examples:
  # Simple deployment. Only dev image, aka ingress and executor will be deployed to cluster
  sh deploy.sh

  # Start Kind cluster and deploy everything on it with local build of images
  sh deploy.sh -a -l -s
```

```bash
$ ./dev/bin/undeploy.sh -h
Usage: undeploy.sh [OPTIONS]

Options:
  -s                  Stop the cluster

Examples:
  # Just deploy all resources from cluster
  sh undeploy.sh

  # Undeploy and stop cluster
  sh undeploy.sh -s
```

_Running integration-tests locally would be done in 2 commands:_
```
$ ./dev/bin/deploy.sh -l -c minikube -s -a
$ ./integration-tests/run-local-tests.sh
```

**Detailed modifications:**
- [x] support for Kind cluster
- [x] created `dev/config` folder and moved there `dev/bin/localconfig` and `dev/bin/environment`, that latter one becoming `interactionconfig` (`dev/bin/common.sh` contains automatic migration commands)
- [x] `dev/_deploy` contains all information relative to the local deployment
  - `dev/_deploy/kustomize` contains a copy of `kustomize` folder with base and  `minikube` overlay. All link/modifications are then done in this folder
  - `dev/_deploy/local_env` contains information about the current local deployment (cluster, cluster_ip, manager_url, keycloak_url, interaction_ob_token if any)
- [x] `dev/bin/configure.sh` renamed to `dev/bin/common.sh`
  - added some `configure_*` methods. scripts don't source with configuration profiles, we call directly the `config`
  - contains useful dirs
  - source in all other script files
  - other methods to read/write to `dev/deploy/local_env` file
  - `get_keycloak_access_token` refactored to handle getting token from internal keycloak url. Useful with Kind as localhost is not accessible from pods (can only reference it external with `localhost`, no external IP ... )
  - `get_sso_access_token` added to retrieve token from Red Hat SSO
- [x] added `deploy.sh` and `undeploy.sh` scripts
- [x] renamed other `dev/bin/*` scripts for more coherence
- [x] refactored a bit the `dev/bin/interaction` scripts to reuse methods from `dev/bin/common.sh` scripts and avoid to get a token for each call. Configuration is now done via `dev/config/interaction-config` file
- [x] updated docs

Please make sure that your PR meets the following requirements:

- [ ] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config)
- [ ] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [ ] All new functionality is tested
- [ ] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

* <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

* <b>Rebase the pull request</b>  
  Comment with: `/rebase`.

</details>

kostola commented on dev/README.md:
```diff
@@ -42,11 +42,11 @@ Here is a list of the **required** environment variable:
 ### Configuration via "localconfig" file
 
-If a file named `localconfig` is created in the [dev/bin](bin) folder, all its content is loaded before every script.
+If a file named `localconfig` is created in the [dev/config](bin) folder, all its content is loaded before every script.
```
you should change the name of the directory also in the link destination, otherwise it will be inconsistent (`[dev/config](config)`)

radtriste replied:
I need to update the whole doc with this new script :)

radtriste replied:
doc updated

kostola commented on dev/bin/configure.sh:
```diff
@@ -93,12 +118,23 @@ function configure_minikube {
+  ping -c 1 "${CLUSTER_IP}" || die "kind is not responding to ping. Is it started?"
 }
 
 function getKeycloakAccessToken {
```
I think we should uniform the function names to snake case instead of mixing it with camel case

radtriste replied:
you are right. gonna update those methods

radtriste replied:
done

kostola commented on dev/bin/configure.sh:
```diff
@@ -37,17 +37,36 @@ for tool in $required_tools; do
 done
 
+# export useful paths
+export ROOT_DIR=`realpath ${CONFIGURE_SCRIPT_DIR_PATH}/../..`
+
+export DEV_DIR="${ROOT_DIR}/dev"
+export LOCAL_DEPLOY_DIR="${DEV_DIR}/_deploy"
+export DEV_CONFIG_DIR="${DEV_DIR}/config"
+export DEV_BIN_DIR="${DEV_DIR}/bin"
+export DEV_LOGS_DIR="${DEV_DIR}/logs"
+export DEV_DOCKER_COMPOSE_DIR="${DEV_DIR}/docker-compose"
+export CREDENTIALS_DIR="${DEV_DIR}/credentials"
+export KUSTOMIZE_DEPLOY_DIR="${LOCAL_DEPLOY_DIR}/kustomize"
+
+export LOCAL_ENV_FILE="${LOCAL_DEPLOY_DIR}/local_env"
+export LOCAL_CONFIG_FILE="${DEV_CONFIG_DIR}/localconfig"
+
+export KUSTOMIZE_DIR="${ROOT_DIR}/kustomize"
```
Why are we exporting all these variables?

kostola replied:
Also, we should use lowercase names for local variables and uppercase only for environment variables to avoid conflicts (see https://stackoverflow.com/a/673940)

radtriste replied:
those are meant to be used in other scripts. so they should be considered as env vars. that is why we are exporting them

kostola replied:
But all other scripts are sourcing `configure.sh`, so these variables will be already set without exporting them (which increases the risk of conflicts with existing env variables).

I would use uppercase variables only for actual env variables that the user can use to configure the scripts (e.g. `MANAGED_KAFKA_INSTANCE_NAME`).

radtriste replied:
I updated all of those. we now use only lowercase names for variables coming from `configure.sh` (renamed to `common.sh`)

danielezonca commented on .github/actions/kind/action.yml:
```diff
@@ -7,16 +7,16 @@ runs:
       shell: bash
       run: |
-        kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/controller-v1.0.4/deploy/static/provider/kind/deploy.yaml
+        kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/main/deploy/static/provider/kind/deploy.yaml
```
Why this change? I would prefer to keep the version fixed to reduce CI instability

radtriste replied:
> Just one minor comment, I'm not doing a proper review. My only consideration is that it looks like our scripting "toolkit" is increasing fast so we might consider to cover them with some tests and/or move to a different language/framework that is easier to test (i.e. jbang/picocli but there are many of them). @robpblake Wdyt?

I do see your point. Right now I just aggregated what is currently existing in an "orchestration" script. Not much new.
I guess we could later migrate to jbang or other

radtriste replied:
fallback done

danielezonca reviewed (commented):
Just one minor comment, I'm not doing a proper review.
My only consideration is that it looks like our scripting "toolkit" is increasing fast so we might consider to cover them with some tests and/or move to a different language/framework that is easier to test (i.e. jbang/picocli but there are many of them).
@robpblake 
Wdyt?

radtriste commented:
> Just one minor comment, I'm not doing a proper review. My only consideration is that it looks like our scripting "toolkit" is increasing fast so we might consider to cover them with some tests and/or move to a different language/framework that is easier to test (i.e. jbang/picocli but there are many of them). @robpblake Wdyt?

I do see your point. Right now I just aggregated what is currently existing in an "orchestration" script. Not much new.
I guess we could later migrate to jbang or other

radtriste commented:
for testing the shell script, we can use the [bats testing](https://opensource.com/article/19/2/testing-bash-bats) 
I can add some tests to this PR

danielezonca commented on dev/bin/common.sh:
```diff
@@ -0,0 +1,229 @@
+  keycloak_token_offline_access=${KEYCLOAK_TOKEN_OFFLINE_ACCESS:-true}
+}
+
+function configure_manager {
```
In many scripts I see the usage of `configure_manager_url`. Is it the same? (and btw, where is defined?)

radtriste replied:
gonna check

danielezonca commented on dev/README.md:
```diff
@@ -149,7 +321,7 @@ and to check the logs, run:
-## Start the Fleet Manager
+### Start the Fleet Manager
 
 The [manager-run.sh](bin/manager-run.sh) script starts the Fleet Manager with all the configurations.
```
To be updated? I don't think the script is named `manager-run.sh` anymore (same comment for the other links)

danielezonca commented on dev/bin/common.sh:
```diff
@@ -0,0 +1,229 @@
+  minikube_profile="${MINIKUBE_PROFILE:-minikube}"
+  minikube_cpus="${MINIKUBE_CPUS:-4}"
+  minikube_memory="${MINIKUBE_MEMORY:-8192}"
+  minikube_kubernetes_version="${MINIKUBE_KUBERNETES_VERSION:-v1.21.6}"
```
This should be the same defined in the .github/actions/kind/action.yml 
Can you add a comment to keep them aligned?

radtriste replied:
I was more thinking about an `.env` file in the root of the project. wdyt ?

danielezonca replied:
Even better :)

radtriste replied:
implemented. please have a look. I had to update the `kind` github action

danielezonca commented on dev/bin/common.sh:
```diff
@@ -0,0 +1,229 @@
+  kind_config_file="${KIND_CONFIG_FILE:-"${dev_config_dir}/kind/config.yaml"}"
+  kind_container_engine="${KIND_CONTAINER_ENGINE}"
+  kind_network="${KIND_NETWORK}"
+  kind_kubernetes_version="${KIND_KUBERNETES_VERSION:-v1.21.10}"
```
Why `1.21.10` while minikube has `1.21.6`?

danielezonca commented on dev/bin/cluster-stop.sh:
```diff
@@ -0,0 +1,27 @@
+
+configure_cluster
+
+if [ "${cluster_type}" = 'minikube' ]; then
+    echo "Stopping Minikube"    
+    ${dev_bin_dir}/cluster-minikube-stop.sh
+elif [ "${cluster_type}" = 'kind' ]; then
+    echo "Stopping Kind"
+    ${dev_bin_dir}/cluster-kind-stop.sh
+else
```
This is not strictly related with this PR but I think we should decide if we test kind or minikube for local development and align CI, current kind + minikube scripting situation has an overhead that we can probably get rid of.

@robpblake 
Wdyt?

robpblake replied:
@danielezonca Yes agreed. A single approach with either KIND or Minikube makes the most sense. I don't have a strong opinion either way - both should work just fine - but we don't have capacity to support both.

jomarko replied:
+1 to point, the amount of introduced scripts increases the maintain costs. Assuming @radtriste is moving to other project after this is finished. I vote for having single option.

danielezonca commented on dev/bin/cluster-start.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash -e
+
+########
+# Create and configure a new minikube cluster
+#
+# Env vars:
+# - KIND_NAME: set the current minikube profile (optional, default none)
+# - KIND_CONFIG_FILE: kind config file (https://kind.sigs.k8s.io/docs/user/quick-start/#configuring-your-kind-cluster) (optional. Default is `dev/config/kind/config.yaml`)
+# - KIND_CONTAINER_ENGINE: container engine to use (optional, default is docker)
+# - KIND_NETWORK: If you want to use a different network from the default Kind one (optional)
+# - KIND_KUBERNETES_VERSION: Kubernetes version to use (optional, default="v1.20.0")
```
To be updated?

danielezonca commented on dev/bin/cluster-stop.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash -e
+
+########
+# Create and configure a new minikube cluster
+#
+# Env vars:
+# - KIND_NAME: set the current minikube profile (optional, default none)
+# - KIND_CONFIG_FILE: kind config file (https://kind.sigs.k8s.io/docs/user/quick-start/#configuring-your-kind-cluster) (optional. Default is `dev/config/kind/config.yaml`)
+# - KIND_CONTAINER_ENGINE: container engine to use (optional, default is docker)
+# - KIND_NETWORK: If you want to use a different network from the default Kind one (optional)
+# - KIND_KUBERNETES_VERSION: Kubernetes version to use (optional, default="v1.20.0")
```
To be updated

danielezonca commented on dev/bin/cluster-kind-stop.sh:
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash -e
+
+########
+# Create and configure a new kind cluster
```
To be updated

danielezonca commented on dev/bin/undeploy.sh:
```diff
@@ -0,0 +1,60 @@
+kustomize build ${kustomize_deploy_dir}/overlays/minikube | kubectl delete -f - || true
+
+echo "---------------- Remove remaining OB namespaces ----------------"
+# This could be replaced by a read in manager of all OB instances running and wait for all of them to be deleted before deleting the resources
```
Yes I agree with this change, current grep based on `ob-` prefix is error prone: that prefix is configurable

radtriste commented on dev/README.md:
```diff
@@ -1,12 +1,14 @@
 You will need the following installed locally on your machine to support local development:
 
-* [Minikube v1.16.0](https://minikube.sigs.k8s.io/docs/start/)
+* [Minikube v1.25.2](https://minikube.sigs.k8s.io/docs/start/) or [Kind](https://kind.sigs.k8s.io/docs/user/quick-start)
```
gonna add it

radtriste commented on dev/README.md:
```diff
@@ -1,12 +1,14 @@
 You will need the following installed locally on your machine to support local development:
 
-* [Minikube v1.16.0](https://minikube.sigs.k8s.io/docs/start/)
+* [Minikube v1.25.2](https://minikube.sigs.k8s.io/docs/start/) or [Kind](https://kind.sigs.k8s.io/docs/user/quick-start)
```
don't know how much time my PR will take to be merged. so better to merge the one from @lucamolteni first and will rebase mine then

lucamolteni replied:
Yes rebase will fix this

radtriste commented on dev/bin/localconfig:
```diff
@@ -3,11 +3,12 @@
-MANAGED_KAFKA_INSTANCE_NAME=mk-local
-OPENSHIFT_OFFLINE_TOKEN=my-offline-token
+managed_connectors_cluster_id=tradisso-cluster-id
+MANAGED_KAFKA_INSTANCE_NAME=tradisso-ob
+OPENSHIFT_OFFLINE_TOKEN=eyJhbGciOiJIUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJhZDUyMjdhMy1iY2ZkLTRjZjAtYTdiNi0zOTk4MzVhMDg1NjYifQ
+# RHOAS_USERNAME=
 
 # optional
 # MANAGED_KAFKA_REGION=
-# MINIKUBE_DRIVER=podman
+MINIKUBE_DRIVER=podman
 # MINIKUBE_CONTAINER_RUNTIME=cri-o
```
oops. will remove

radtriste commented on dev/bin/local-manager-stop.sh:
```diff
@@ -0,0 +1,16 @@
+. $(dirname "${BASH_SOURCE[0]}")/common.sh
+
+echo "Kill local manager process(es)"
+kill_process_with_name 'Dquarkus.http.port=8080'
```
I see what you mean. gonna merge into a `local-service-stop.sh` file

radtriste replied:
@danielezonca I updated the stop scripts. Can you please have a look ?

radtriste replied:
it will now look to the `target/*-dev.jar` file and do the kill based on that file.

radtriste commented on dev/bin/interaction/get-all-bridges.sh:
```diff
@@ -2,4 +2,9 @@
 
-curl -s -H "Authorization: $OB_TOKEN" -X GET $MANAGER_URL/api/v1/bridges | jq .
+check_token
+configure_manager
+
+echo ${manager_url}
```
gonna verify that

radtriste commented on dev/bin/interaction/create-processor.sh:
```diff
@@ -2,15 +2,18 @@
 
 source "$(dirname "${BASH_SOURCE[0]}")/common.sh"
```
could the problem of rhoas CLI be that it targets automatically the console.redhat.com URL and not the local one ?

danielezonca replied:
No currently we don't have CLI support for OB in rhoas but I think we should do it ~soon so that we can use it directly

jomarko replied:
@danielezonca is such task already in jira or available to work on? I mean the rhoas CLI commands for smart events.

lucamolteni commented:
if we decide to go for this, I think we should communicate it to the team properly as the structure of the scripts seems to have changed significantly

radtriste commented:
> if we decide to go for this, I think we should communicate it to the team properly as the structure of the scripts seems to have changed significantly

Yes that was my intention :)
once merged, I will write on #dev room

manstis commented:
> once merged, I will write on #dev room

I prefer emails to be honest, pings on Google Chat are easily lost (or it's just me).

danielezonca commented on dev/bin/logs-shard.sh:
```diff
@@ -0,0 +1,25 @@
+    if [ -z ${pod_name} ]; then
+        echo 'Cannot find the manager pod...'
+        echo
+        oc get pods -n event-bridge-operator
```
Please convert to kubectl or add oc to required tools

radtriste replied:
woops ...

radtriste commented:
- rebased
- migration should be easier
  - added migration guide in dev README file
  - config file migration will be done automatically

@lucamolteni @manstis @danielezonca @sutaakar please have a look

sutaakar commented on dev/README.md:
```diff
@@ -17,46 +19,162 @@ You will need the following installed locally on your machine to support local d
 
-Do not install Minikube via `brew`. 
-Download the specific v.1.16.0 from the [Minikube release page](https://github.com/kubernetes/minikube/releases/tag/v1.16.0).
+Configuration can be done via environment variables or via the `dev/bin/localconfig` file.
```
```suggestion
Configuration can be done via environment variables or via the `dev/config/localconfig` file.
```

sutaakar commented on dev/README.md:
```diff
@@ -17,46 +19,162 @@ You will need the following installed locally on your machine to support local d
+
+TODO minikube / kind
+
+#### Minikube cofniguration for macOS users
```
```suggestion
#### Minikube configuration for macOS users
```

sutaakar commented on dev/README.md:
```diff
@@ -17,46 +19,162 @@ You will need the following installed locally on your machine to support local d
+
+### Local cluster configuration
+
+TODO minikube / kind
```
What should this refer to?

radtriste replied:
I just forgot to remove the TODO ...

```suggestion
```

sutaakar commented on dev/README.md:
```diff
@@ -17,46 +19,162 @@ You will need the following installed locally on your machine to support local d
+You will find below some explanations of those.
 
-## Managed Kafka instance setup
+### Setup Kafka
```
```suggestion
### Setup Managed Kafka
```

radtriste commented on dev/README.md:
```diff
@@ -17,46 +19,162 @@ You will need the following installed locally on your machine to support local d
+
+### Local cluster configuration
+
+TODO minikube / kind
+
```
```suggestion
```

radtriste commented on dev/README.md:
```diff
@@ -17,46 +19,162 @@ You will need the following installed locally on your machine to support local d
+
+* start the configured cluster (`-s` option)
+* build the local image and push them to cluster (`-l` option)
+* setup kafka resources on managed services (`-k` option)
```
true... forgot to remove it after the rebase

radtriste replied:
```suggestion
```

r00ta reviewed (commented):
That's very good job @radtriste , but I have some concerns about the maintenance of those scripts since you are going to leave the project. 
Personally, I'm already struggling when I have to update the scripts for small changes because they require some knowledge, and if we put more and more overhead in the scripts it is going to be even worse. 

Don't get me wrong, I know there is a lot of effort in this but I'm not sure this is the right thing to do now.

radtriste commented:
Honestly, I just reused what is existing (with a few renaming) and added a "flow" script to run one after the other, as described in the documentation.
I did not add any magic... The complexity did not changed.

But if you feel like it should not be merged, then don't. I am also ok with that.

jomarko commented on dev/README.md:
```diff
@@ -179,3 +340,26 @@ The script runs forever, on linux machines press `CTRL+C` to stop it (or just se
+
+## Config files moved
+
+On the first execution of the `bin/common.sh` script, configuration files will be automatically transferred to the `config` folder.  
+Here are the moves done:
```
What is the recommended way for this migration. In past , I have never run `bin/common.sh` on its own. Is it the same? If not, we should explicitly mention, that it is first step of migration.

radtriste replied:
this script is a utils script which is called from all other scripts. You should not need to run it.
Here is maybe a better sentence. Let me know if you are ok with that @jomarko 

```suggestion
On the execution of any script, it should call the `bin/common.sh` script, which will automatically transfer the configuration files to the `config` folder.  
Here are the moves done:
```

jomarko replied:
that sounds better to me @radtriste , yes

jomarko reviewed (commented):
Thank you for the PR @radtriste
